consensus.go

```diff
@@ -0,0 +1,240 @@
+package dcrlibwallet
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"sort"
+	"time"
+
+	"decred.org/dcrwallet/v2/errors"
+	w "decred.org/dcrwallet/v2/wallet"
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	"github.com/decred/dcrd/chaincfg/v3"
+)
+
+type Consensus struct {
+	mwRef *MultiWallet
+	ctx   context.Context
+}
+
+func newConsensus(mwRef *MultiWallet) (*Consensus, error) {
+	c := &Consensus{
+		mwRef: mwRef,
+	}
+
+	return c, nil
+}
+
+// GetVoteChoices handles a getvotechoices request by returning configured vote
+// preferences for each agenda of the latest supported stake version for the
+// selected wallet.
+func (c *Consensus) GetVoteChoices(ctx context.Context, hash string, walletID int) (*GetVoteChoicesResult, error) {
+	wallet := c.mwRef.WalletWithID(walletID)
+	wal := wallet.Internal()
+	if wal == nil {
+		return nil, fmt.Errorf("request requires a wallet but wallet has not loaded yet")
+	}
+
+	var ticketHash *chainhash.Hash
+	if hash != "" {
+		hash, err := chainhash.NewHashFromStr(hash)
+		if err != nil {
+			return nil, fmt.Errorf("inavlid hash: %w", err)
+		}
+		ticketHash = hash
+	}
+
+	version, agendas := getAllAgendas(wal.ChainParams())
+	resp := &GetVoteChoicesResult{
+		Version: version,
+		Choices: make([]VoteChoice, len(agendas)),
+	}
+
+	choices, _, err := wal.AgendaChoices(ctx, ticketHash)
+	if err != nil {
+		return nil, err
+	}
+
+	for i := range choices {
+		resp.Choices[i] = VoteChoice{
+			AgendaID:          choices[i].AgendaID,
+			AgendaDescription: agendas[i].Vote.Description,
+			ChoiceID:          choices[i].ChoiceID,
+			ChoiceDescription: "", // Set below
+		}
+		for j := range agendas[i].Vote.Choices {
+			if choices[i].ChoiceID == agendas[i].Vote.Choices[j].Id {
+				resp.Choices[i].ChoiceDescription = agendas[i].Vote.Choices[j].Description
+				break
+			}
+		}
+	}
+
+	return resp, nil
+}
+
+// SetVoteChoice handles a setvotechoice request by modifying the preferred
+// choice for a voting agenda.
+//
+// If a VSP host is configured in the application settings, the voting
+// preferences will also be set with the VSP.
+func (c *Consensus) SetVoteChoice(walletID int, vspPubKey []byte, vspHost, agendaID, choiceID, hash, passphrase string) error {
```

Comment:
Comment needs some love.
`passphrase` should be `[]byte`.
`vspHost string`, before `vspPubKey []byte`.

Reply:
comment is same as upstream, but i can still modify it

Reply:
Yeah, the comment upstream makes more sense because there is an application setting that stores a pre-configured vsp host and also because the method is actually a handler for an rpc request. This one isn't an handler in that sense but does similarly.